servant-server/src/Servant/Server/Internal/RoutingApplication.hs

```diff
@@ -112,12 +115,33 @@ instance Functor (Delayed env) where
       , ..
       } -- Note [Existential Record Update]
 
+-- | A mutable cleanup action
+newtype CleanupRef = CleanupRef (IORef (IO ()))
+
+newCleanupRef :: IO CleanupRef
+newCleanupRef = CleanupRef <$> newIORef (return ())
+
+-- | Add a clean up action to a 'CleanupRef'
+addCleanup' :: IO () -> CleanupRef -> IO ()
```

Comment:
I'd just have this is a `where` clause to `addCleanup` so it's obvious that it's not meant to be used elsewhere.